### ISSUE TITLE
undefined method `tag_name` for shadow root

### ISSUE THREAD

**Issue by eric-hemasystems** [teamcapybara/capybara#2586]
## Meta
Capybara Version:
3.37.1

Driver Information (and browser if relevant):
selenium-webdriver 4.3.0
Chrome 105.0.5195.125

## Expected Behavior

When a shadow root is used it should not cause an error. Example code:

```ruby
within(browse_rooms_list.shadow_root) do
  expect(page).to have_css('bc-room-list-item', count: room_count)
end
```

## Actual Behavior

```
undefined method `tag_name' for #<Selenium::WebDriver::ShadowRoot:0x00007feab1161bd0> (NoMethodError)
      /Users/eric/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/capybara-3.37.1/lib/capybara/selenium/node.rb:187:in `tag_name'
      /Users/eric/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/capybara-3.37.1/lib/capybara/node/element.rb:578:in `inspect'
      /Users/eric/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/capybara-3.37.1/lib/capybara/queries/selector_query.rb:105:in `description'
      /Users/eric/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/capybara-3.37.1/lib/capybara/queries/selector_query.rb:114:in `applied_description'
      /Users/eric/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/capybara-3.37.1/lib/capybara/queries/selector_query.rb:179:in `failure_message'
      /Users/eric/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/capybara-3.37.1/lib/capybara/result.rb:114:in `failure_message'
      /Users/eric/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/capybara-3.37.1/lib/capybara/node/matchers.rb:112:in `block in assert_selector'
      /Users/eric/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/capybara-3.37.1/lib/capybara/node/matchers.rb:844:in `block in _verify_selector_result'
      /Users/eric/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/capybara-3.37.1/lib/capybara/node/base.rb:83:in `synchronize'
      /Users/eric/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/capybara-3.37.1/lib/capybara/node/matchers.rb:843:in `_verify_selector_result'
      /Users/eric/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/capybara-3.37.1/lib/capybara/node/matchers.rb:110:in `assert_selector'
      /Users/eric/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/capybara-3.37.1/lib/capybara/session.rb:771:in `assert_selector'
      /Users/eric/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/capybara-3.37.1/lib/capybara/rspec/matchers/have_selector.rb:18:in `element_matches?'
      /Users/eric/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/capybara-3.37.1/lib/capybara/rspec/matchers/base.rb:51:in `matches?'
      /Users/eric/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/rspec-expectations-3.10.1/lib/rspec/expectations/handler.rb:51:in `block in handle_matcher'
      /Users/eric/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/rspec-expectations-3.10.1/lib/rspec/expectations/handler.rb:27:in `with_matcher'
      /Users/eric/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/rspec-expectations-3.10.1/lib/rspec/expectations/handler.rb:48:in `handle_matcher'
      /Users/eric/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/rspec-expectations-3.10.1/lib/rspec/expectations/expectation_target.rb:65:in `to'
...
```

## Steps to reproduce

Seems to do it even when passing despite the backtrace indicating it is building the failure method. Perhaps it some sort of wait/retry mechanism built into Capybara that means the failure message is being generated?

The following monkey-patch seems to solve:

```ruby
Selenium::WebDriver::ShadowRoot.class_eval do
  def tag_name
    @id
  end
end
```

I don't know if the @id is the right thing to return or perhaps `Capybara::Selenium::Node` needs to be adjusted to not even call `tag_name` if the native element doesn't define it?

**Comment by twalpole** [comment]
Not sure what the right thing to return here is -- will hopefully have time to look at it this week --- Note this should only be occurring if your assertion is failing, so if your tests are correctly passing you shouldn't see this error.

**Comment by twalpole** [comment]
fixed via - add69ef8a21fc96a588d0338d810901c0aa41298

**Comment by twalpole** [comment]
@eric-hemasystems leftover debug code that happily didn't actually hurt anything (ambiguous interpretation).  Thanks for pointing it out, it's been removed.
